rotkehlchen/chain/ethereum/modules/lido/decoder.py

```diff
@@ -0,0 +1,526 @@
+                (
+                    (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.RECEIVE_WRAPPED) or  # noqa: E501
+                    (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.RETURN_WRAPPED)  # noqa: E501
+                )
+            ):
+                event.notes = event.notes.format(amount=event.balance.amount)  # set the amount  # noqa: E501
+                break
+        return decoded_events
+
+    def _decode_lido_transfer_shares_for_steth_wrap(
+            self, context: DecoderContext,
+    ) -> DecodingOutput:
+        """Decode steth wrapping in wsteth from TransferShares log event"""
+
+        user_address = hex_or_bytes_to_address(context.tx_log.topics[1])
+
+        #  we need to retrieve the transferred stETH amount deterministically.
+        #  expected log index for the transferred steth is -1
+        #  expected log index for the transferred wsteth is -3
+        expected_steth_transfer_index = context.tx_log.log_index - 1
+        expected_wsteth_transfer_index = context.tx_log.log_index - 3
+        paired_wsteth_ratio = None
+        sth_wrapped_amount = None
+        wsteth_received_amount = None
+
+        for tx_log in context.all_logs:
+            if (
+                tx_log.log_index == expected_steth_transfer_index and
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                tx_log.address == steth_evm_address and
+                hex_or_bytes_to_address(tx_log.topics[1]) == user_address and  # from
+                hex_or_bytes_to_address(tx_log.topics[2]) == wsteth_evm_address  # to # wsteth
+            ):
+                sth_wrapped_amount = asset_normalized_value(
+                    amount=hex_or_bytes_to_int(tx_log.data[:32]),
+                    asset=A_STETH.resolve_to_crypto_asset(),
+                )
+            elif (
+                tx_log.log_index == expected_wsteth_transfer_index and
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                tx_log.address == wsteth_evm_address and
+                hex_or_bytes_to_address(tx_log.topics[1]) == ZERO_ADDRESS and  # from
+                hex_or_bytes_to_address(tx_log.topics[2]) == user_address  # to # wsteth
+            ):
+                wsteth_received_amount = asset_normalized_value(
+                    amount=hex_or_bytes_to_int(tx_log.data[:32]),
+                    asset=A_WSTETH.resolve_to_crypto_asset(),
+                )
+
+            if sth_wrapped_amount and wsteth_received_amount:
+                paired_wsteth_ratio = FVal(sth_wrapped_amount) / FVal(wsteth_received_amount)
+                break
+
+        if paired_wsteth_ratio is None:
+            return DEFAULT_DECODING_OUTPUT
+
+        in_event = None
+        out_event = None
+        for event in context.decoded_events:
```

Comment:
okay so this function has gotten super complicated. Iterating all log entries. And then all decoded events. I do not understand what you are trying to achieve. It seems too complicated and perhaps you are not understanding how the decoders work.